// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"placio-app/ent/fitness"
	"placio-app/ent/predicate"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// FitnessDelete is the builder for deleting a Fitness entity.
type FitnessDelete struct {
	config
	hooks    []Hook
	mutation *FitnessMutation
}

// Where appends a list predicates to the FitnessDelete builder.
func (fd *FitnessDelete) Where(ps ...predicate.Fitness) *FitnessDelete {
	fd.mutation.Where(ps...)
	return fd
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (fd *FitnessDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, fd.sqlExec, fd.mutation, fd.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (fd *FitnessDelete) ExecX(ctx context.Context) int {
	n, err := fd.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (fd *FitnessDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(fitness.Table, sqlgraph.NewFieldSpec(fitness.FieldID, field.TypeString))
	if ps := fd.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, fd.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	fd.mutation.done = true
	return affected, err
}

// FitnessDeleteOne is the builder for deleting a single Fitness entity.
type FitnessDeleteOne struct {
	fd *FitnessDelete
}

// Where appends a list predicates to the FitnessDelete builder.
func (fdo *FitnessDeleteOne) Where(ps ...predicate.Fitness) *FitnessDeleteOne {
	fdo.fd.mutation.Where(ps...)
	return fdo
}

// Exec executes the deletion query.
func (fdo *FitnessDeleteOne) Exec(ctx context.Context) error {
	n, err := fdo.fd.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{fitness.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (fdo *FitnessDeleteOne) ExecX(ctx context.Context) {
	if err := fdo.Exec(ctx); err != nil {
		panic(err)
	}
}
